modules/message-lane/src/lib.rs

```diff
@@ -90,19 +159,237 @@ decl_module! {
 		/// Deposit one of this module's events by using the default implementation.
 		fn deposit_event() = default;
 
+		/// Block initialization.
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let mut dispatcher = T::QueuedMessageDispatch::with_any_weight();
+			let (queued_messages_dispatched, weight_spent_on_queued_dispatch) = dispatcher.dispatch();
+			if queued_messages_dispatched != 0 {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Spent {} weight on {} queued messages from on_initialize()",
+					weight_spent_on_queued_dispatch,
+					queued_messages_dispatched,
+				);
+			}
+
+			weight_spent_on_queued_dispatch
+		}
+
 		/// Send message over lane.
 		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
 		pub fn send_message(
 			origin,
 			lane_id: LaneId,
 			payload: T::Payload,
-		) {
-			let _ = ensure_signed(origin)?;
+			delivery_and_dispatch_fee: T::MessageFee,
```

Comment:
I feel this should rather be calculated by `LaneMessageVerifier` - it's supposed to verify that so it knows the minimal value anyway. Is it possible to create a market for messages (i.e. higher paying fees go first?). Maybe we should change that to just `tip`, and have the sender pay the fee calculated by `LaneMessageVerifier` plus the tip?

Reply:
Since the target chain congestion might be dynamic, I imagine we will require some extra way to increase the `tip` for a specific message, to make sure that relayers are either waiting for a less-congested time (stalling the delivery) or overpay for transaction (lose money). I can see that being a completely separate PR, where you just add an extra "unlocking lane reward". I.e. the senders that see their messages (their lane) being stuck due to insufficient fees can just top-up the entire lane.

Reply:
> it's supposed to verify that so it knows the minimal value anyway

Hmm...It **may** verify that. But I don't think that all implementations must do that. I suggested one possible way on how to estimate `source-fees -> target-weight` conversions in the description. But it isn't required and needs to be confirmed yet. 

Imo we can't rely on the fact that **any** source chain knows **current state** of target chain economics (i.e. conversion_rate(source -> target), `TargetWeightToFee(TargetWeight)`). I'm going to hardcode this knowledge for our test nets (simply 1:1), but for other chains (Kusama, Polkadot) we can't rely on that.

My vision is: (1) sender selects `delivery_and_dispatch_fee` on its own, probably we could add some RPCs to relay node (that will has access to both chains + probably market information); (2) relay just ignores messages/lanes with too low `delivery_and_dispatch_fee`.

> I imagine we will require some extra way to increase the tip

I thought we have agreed that this could be solved just by pushing additional message to the lane with some dummy call (probably just empty vec?)? The fee should be enough to cover both this and previous message(s) costs.

Reply:
> Hmm...It may verify that.

AFAICT, if it doesn't verify then it's broken. I agree that it's cool to have it extensible, but at the same time too much extensibility will just let the users (or even the future versions of us) shoot themselves in the foot. Added a more detailed comment on this below.

> Imo we can't rely on the fact that any source chain knows current state of target chain economics (i.e. conversion_rate(source -> target), TargetWeightToFee(TargetWeight)). 

Hmm, that's a fair point, but IMHO it needs at least some minimal knowledge, that might be slightly out of date. Obviously we will never be able to synchronize the knowledge fast enough, to respond to the changing market, but imo it's viable to assume we know some long term trend at least.

That said, the fees would be worked on in a separate PR anyway, when we do proper integration with the runtime and I'm happy to pause the discussion here and resume in this other PR. Referencing https://github.com/paritytech/parity-bridges-common/issues/316 here.

Reply:
As I've been writing in all other comments, I disagree with your points on hardcoding/decoupling logic. But I see no benefits in continuing this discussion, because it isn't progressing at all && we're just wasting time here. I'll turn on agnostic mode for this kind of comments here && apply your suggestions.

Reply:
Let's leave it as-is now, I need to sleep over this again, cause I think I'm starting to see your concerns, and re-review after we removed the queue.

Reply:
Fine to leave it as-is, we can revisit later.